### ISSUE TITLE
release-20.1: tables with indexes with prefixes with all columns of an fk can lead to bad cascades

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#54547]
**Describe the problem**

The bug described here could lead to columns not involved in a foreign key relationship to be erroneously updated when perfoming a cascaded update due to `ON DELETE SET DEFAULT` or `ON DELETE SET NULL`. The end result of this bug was that columns could be updated with values that might violate `CHECK` or `NOT NULL` constraints.

**To Reproduce**

To reproduce one needs to create a table with a foreign key and also have an index which contains the columns of the foreign key plus additional columns that occurs earlier in the index list.

```sql
CREATE TABLE referenced (
    j INT PRIMARY KEY
);

CREATE TABLE origin (
    i INT PRIMARY KEY,
    j INT,
    k INT NOT NULL DEFAULT -1,
    INDEX (j, k),
    INDEX (j)
);
```

Then add the constraint (with either `SET DEFAULT` or `SET NULL`):
```sql
ALTER TABLE origin
    ADD CONSTRAINT fk FOREIGN KEY (j) REFERENCES referenced(j) ON DELETE SET DEFAULT;
```
or
```sql
ALTER TABLE origin
    ADD CONSTRAINT fk FOREIGN KEY (j) REFERENCES referenced(j) ON DELETE SET NULL;
```

Add some rows:
```sql
INSERT INTO referenced VALUES (1);
INSERT INTO origin VALUES (1, 1, 1);
```

Delete the referenced row:
```sql
DELETE FROM referenced WHERE j = 1;
```

Prior to #54543 this would lead to `k` being updated in the cascade (which it should not be):

```sql
SELECT * FROM origin;
  i |  j   | k
----+------+-----
  1 | NULL | -1
(1 row)

```

Or with the `SET NULL` case, we'd actually write `NULL` into the `NOT NULL` column:

```
SELECT * FROM fk_54208_origin_null;
ERROR: internal error: Non-nullable column "origin:k" with no value! Index scanned was "primary" with the index key columns (i) and the values (1)
SQLSTATE: XX000
DETAIL: stack trace:
/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:1440: finalizeRow()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:1176: NextRow()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:217: Next()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:749: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:370: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:405: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1019: PlanAndRun()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:938: execWithDistSQLEngine()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:831: dispatchToExecutionEngine()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:536: execStmtInOpenState()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:96: execStmt()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1406: execCmd()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1335: run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:479: ServeConn()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:582: func1()
runtime/asm_amd64.s:1357: goexit()

HINT: You have encountered an unexpected error.

Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.

If you would rather not post publicly, please contact us directly
using the support form.

We appreciate your feedback.
```

**Expected behavior**
Foreign key cascades would work.

**Environment:**
 - CockroachDB version 20.1.5
 - Server OS: [e.g. Linux/Distrib]
 - Client app [e.g. `cockroach sql`, JDBC, ...]

**Comment by yuzefovich** [comment]
@ajwerner should we close this issue?

**Comment by ajwerner** [comment]
Closed by #54543
